airframe-http-finagle/src/main/scala/wvlet/airframe/http/finagle/filter/HttpAccessLogFilter.scala

```diff
@@ -0,0 +1,220 @@
+import wvlet.airframe.http.finagle.filter.HttpAccessLogFilter.{HttpRequestLogger, _}
+import wvlet.log.LogTimestampFormatter
+
+import scala.collection.immutable.ListMap
+import scala.util.control.NonFatal
+
+case class HttpAccessLogFilter(
+    httpAccessLogWriter: HttpAccessLogWriter = HttpAccessLogWriter.default,
+    // Loggers for request contents
+    requestLoggers: Seq[HttpRequestLogger] = defaultRequestLoggers,
+    // Loggers for response contents
+    responseLoggers: Seq[HttpResponseLogger] = defaultResponseLoggers,
+    // Loggers for thread-local storage contents
+    contextLoggers: Seq[HttpContextLogger] = defaultContextLoggers,
+    excludeHeaders: Set[String] = Set(HttpHeader.Authorization, HttpHeader.ProxyAuthorization)
+) extends SimpleFilter[Request, Response] {
+
+  def addRequestLogger(logger: HttpRequestLogger): HttpAccessLogFilter = {
+    this.copy(requestLoggers = requestLoggers :+ logger)
+  }
+  def addResponseLogger(logger: HttpResponseLogger): HttpAccessLogFilter = {
+    this.copy(responseLoggers = responseLoggers :+ logger)
+  }
+  def aadContextLogger(logger: HttpContextLogger): HttpAccessLogFilter = {
+    this.copy(contextLoggers = contextLoggers :+ logger)
+  }
+  def addExcludeHeaders(excludes: Set[String]): HttpAccessLogFilter = {
+    this.copy(excludeHeaders = excludeHeaders ++ excludes)
+  }
+
+  private val sanitizedExcludeHeader = excludeHeaders.map(sanitizeHeader)
+
+  private def emit(m: Map[String, Any]) = {
+    val filtered = m.filterNot(x => sanitizedExcludeHeader.contains(x._1))
+    httpAccessLogWriter.write(filtered)
+  }
+
+  override def apply(request: Request, context: Service[Request, Response]): Future[Response] = {
+    // Use ListMap to preserve the parameter order
+    val m = ListMap.newBuilder[String, Any]
+    for (l <- requestLoggers) {
+      m ++= l(request)
+    }
+
+    val currentNanoTime = System.nanoTime()
+    def millisSince     = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - currentNanoTime)
+
+    def reportError(e: Throwable): Future[Response] = {
+      val responseTimeMillis = millisSince
+      m += "response_time_ms" -> responseTimeMillis
+      reportContext
+      m ++= errorLog(request, e)
+      emit(m.result())
+      Future.exception(e)
+    }
+
+    def reportContext: Unit = {
+      for (l <- contextLoggers) {
+        m ++= l(request)
+      }
+    }
+
+    try {
+      context(request)
+        .map { response =>
+          reportContext
+          m += "response_time_ms" -> millisSince
+          for (l <- responseLoggers) {
+            m ++= l(response)
+          }
+          emit(m.result())
+          response
+        }.rescue {
+          case NonFatal(e: Throwable) =>
+            reportError(e)
+        }
+    } catch {
+      // When an unknown internal error happens
+      case e: Throwable =>
+        reportError(e)
+    }
+  }
+
+  private def errorLog(request: Request, e: Throwable): Map[String, Any] = {
```

Comment:
It might be convenient if we also have `ErrorLogger` interface like:
```scala
type HttpErrorLogger  = (Request, Throwable) => Map[String, Any]
```
Then this `errorLog()` can be a default error logger.

Reply:
Added HttpErrorLogger